### PR TITLE
Qconditions

### PR DESCRIPTION
**Pull request by mbenson**
Additional query conditions for your consideration

### CODE REVIEW

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:248]:
```suggestion
   * @param {string} field string name of compared field
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:257]:
```suggestion
   * @param {string} field string name of compared field
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:280]:
```suggestion
  greaterThanOrEqualToField(field) {
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:300]:
```suggestion
  lessThanOrEqualToField(field) {
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:307]:
```suggestion
   * @param {number} n number of unit
   * @param {TimeUnit} unit of time (year, month, hour, minute)
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:303]:
```suggestion

/**
 * @typedef {"year" | "month" | "day" | "hour" | "minute" | "second"} TimeUnit
 */

```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:317]:
```suggestion
   * @param {number} n number of unit
   * @param {TimeUnit} unit of time (year, month, hour, minute)
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:328]:
```suggestion
   * @param {number} n number of unit
   * @param {TimeUnit} unit of time (year, month, week, day, hour)
   * @returns {RelativeDateBuilder} builder object supporting .before(field) -> original builder
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:338]:
```suggestion
   * @param {number} n number of unit
   * @param {TimeUnit} unit of time (year, month, week, day, hour)
   * @returns {RelativeDateBuilder} builder object supporting .before(field) -> original builder
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:472]:
```suggestion
  /**
   * @param {QueryBuilder} parent Parent builder
   * @param {string} quantifier Relationship between base and n
   * @param {number} n Number of unit
   * @param {TimeUnit} unit Unit of time
   */
  constructor(parent, quantifier, n, unit) {
```

File: query-builder.js
Comment (decompil3d) [inline query-builder.js:482]:
```suggestion
   * @param {string} field to compare against current field
   * @returns {QueryBuilder} parent
```

**Review by decompil3d — CHANGES_REQUESTED** [review]
Thank you for your contribution! I've added some comment suggestions that improve the JSDoc typings on your additions. Can you please add tests for these new methods though?

**Comment by mbenson** [comment]
Hi, thanks for your consideration. I'm relatively new to the js ecosystem; so far I've mostly encountered the pattern where tests are colocated with the files they test and on that basis missed your tests (I wondered why I didn't see any ;) ). Expect them today.

File: query-builder.js
Comment (mbenson) [reply query-builder.js:303]: This is interesting. What I find on running the `build:types` script, however, is that the described type is inlined. Where I documented the applicable types you will notice that they are not all represented in any given location. This reflects my experience testing the various units of time against a SNOW instance (I found the documentation on these conditions to be quite sparse; if you know of a better resource that would be great). If `TimeUnit` would be "promoted" to a named type I would feel okay about it, but if it is to be inlined onto each method then to me that represents a problem for the generated .d.ts to explicitly include a parameter value that, so far as I was able to find, does not work.

Comment (mbenson) [reply query-builder.js:303]: Please consider commit 7735643c2fc2cc9bb3e53ae012003cdabbc913bc .

**Comment by mbenson** [comment]
I've got the unit tests added and am pretty happy with the current state of the code, personally. Perhaps because I've done some squashing, etc., along the way, I can't seem to find where to mark your requested changes as having been completed. :/

File: query-builder.js
Comment (decompil3d) [reply query-builder.js:335]: I trust you on this. I changed jobs and no longer use ServiceNow day to day. So I can't really test this myself.

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:515]:
```suggestion
    });
    it('constructs query properly', function () {
      const res = q.field('foo').greaterThanOrEqualToField('bar').build();
      assume(res).equals('fooGT_OR_EQUALS_FIELDbar');
    });
```

**Review by decompil3d — CHANGES_REQUESTED** [review]
Just one missing test case, otherwise looking good

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:555]:
```suggestion
      // @ts-expect-error
      assume(() => q.since('1', 'hour')).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:558]:
```suggestion
      // @ts-expect-error
      assume(() => q.since(1, true)).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:575]:
```suggestion
      // @ts-expect-error
      assume(() => q.notSince('1', 'hour')).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:578]:
```suggestion
      // @ts-expect-error
      assume(() => q.notSince(1, true)).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:595]:
```suggestion
      // @ts-expect-error
      assume(() => q.isMoreThan('1', 'hour')).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:598]:
```suggestion
      // @ts-expect-error
      assume(() => q.isMoreThan(1, true)).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:615]:
```suggestion
      // @ts-expect-error
      assume(() => q.isLessThan('1', 'hour')).throws(QueryTypeException);
```

File: test/query-builder.test.js
Comment (decompil3d) [inline test/query-builder.test.js:618]:
```suggestion
      // @ts-expect-error
      assume(() => q.isLessThan(1, true)).throws(QueryTypeException);
```

**Review by decompil3d — CHANGES_REQUESTED** [review]
Oh, and some missing ts-expect-error declarations (if you use VSCode with check-JS enabled, you'll see TS errors here)

**Comment by mbenson** [comment]
Thanks again for bearing with me here. IDK why this last rebase dropped you off as a collaborating committer :|

**Comment by mbenson** [comment]
FYI, I have also rebased my `SAMEAS/NSAMEAS` commit to specify that the `field` parameter is a `string`, which allowed me to revert the change I made to `_validateType` that made its `types` parameter optional.

**Review by decompil3d — CHANGES_REQUESTED** [review]
This looks great. Sorry, one last thing -- can you please update the README with the new methods? Once that's done we should be good to merge.

**Review by decompil3d — APPROVED** [review]

**Comment by decompil3d** [comment]
Published in `servicemeow@3.2.0`. Thanks for your contribution!

**Comment by mbenson** [comment]
Thanks for your openness, guidance, and the quick release!

**Pull request merged**
